### PR TITLE
Replace bn.js with BigInt

### PR DESCRIPTION
**Pull request by gtsonevv**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md).
- [x] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [ ] **If this is a code change**: I have written unit tests.
- [x] **If this changes code in a published package**: I have run `pnpm changeset` to create a `changeset` JSON document appropriate for this change.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

Reduce amount of dependencies.

## Related issues/PRs

https://github.com/near/near-api-js/pull/1217
https://github.com/near/near-api-js/issues/1200

### CODE REVIEW

File: packages/accounts/src/account.ts
Comment (gagdiez) [inline packages/accounts/src/account.ts:296]: is this the intended behaviour? this will default to 0 if `nonce` is not defined, was that the case for `new BN(undefined)`? 

This is very relevant, specially since it is inside a `try/catch` statement

File: packages/utils/src/format.ts
Comment (gagdiez) [inline packages/utils/src/format.ts:11]: please check that this handle the right precision, before it was `Bigint(10)**BigInt(EXP)`

File: packages/utils/src/validators.ts
Comment (gagdiez) [inline packages/utils/src/validators.ts:33]: is this right? I would have assumed the operation to be `(left+right)/2`

File: packages/utils/src/validators.ts
Comment (gagdiez) [inline packages/utils/src/validators.ts:24]: `BigInt` have order operators, is it necessary to provide a custom made one?

File: packages/utils/src/validators.ts
Comment (gagdiez) [inline packages/utils/src/validators.ts:56]: same as before

**Review by gagdiez — CHANGES_REQUESTED** [review]
Please see all my comments above

File: packages/utils/src/validators.ts
Comment (gtsonevv) [reply packages/utils/src/validators.ts:24]: this is from the MDN docs.
```
mixed.sort(); // default sorting behavior
// [ -12n, 0, 0n, 10, 4n, 4, 6 ]

mixed.sort((a, b) => a - b);
// won't work since subtraction will not work with mixed types
// TypeError: can't convert BigInt value to Number value

// sort with an appropriate numeric comparator
mixed.sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
// [ -12n, 0, 0n, 4n, 4, 6, 10 ]
```

File: packages/utils/src/format.ts
Comment (gtsonevv) [reply packages/utils/src/format.ts:11]: yes, should be `BigInt(10)**BigInt(EXP)` but we need to change the `target` option to `es2016` or later to make it work. Are you ok with that?

Comment (gagdiez) [reply packages/utils/src/format.ts:11]: yes, since

`BigInt(10**24) == 999999999999999983222784n`

while

`BigInt(10)**BigInt(24) == 1000000000000000000000000n`

**Comment by gagdiez** [comment]
How is that a test is not catching problems here? for example, the change of BN(10).exp(24) to BigInt(10**24) should have broke a bunch of stuff (since it is in the utils.js module, that every other module uses).

Wrongly changing it to `BigInt(10**24) == 999999999999999983222784n` should have broke a lot of tests... or is that we do not have tests for this?

File: packages/accounts/src/account.ts
Comment (gagdiez) [reply packages/accounts/src/account.ts:296]: I tested it myself, and indeed, `new BN(undefined) === BN(0)`

File: packages/utils/src/validators.ts
Comment (gagdiez) [reply packages/utils/src/validators.ts:33]: indeed, this is a bug

```js
const left = new BN(2)
const right = new BN(2)

const mid = left.add(right).div(new BN(2))
```

`mid === 2` while  `left + right / BigInt(2)  === 3n`

how is that this didn't break any test?

File: packages/accounts/src/account.ts
Comment (gtsonevv) [reply packages/accounts/src/account.ts:296]: will leave it `nonce: BigInt(rawAccessKey.nonce || 0)` because `BigInt` constructor doesn't work with `undefined`

File: packages/transactions/src/create_transaction.ts
Comment (vikinatora) [inline packages/transactions/src/create_transaction.ts:8]: `bigIntNonce` is misleading since it can be of type `bigint | string | number`
```suggestion
    const txNonce = typeof nonce === 'string' || typeof nonce === 'number' ? BigInt(nonce) : nonce;
    return new Transaction({ signerId, publicKey, nonce: txNonce, receiverId, actions, blockHash });
```

Comment (vikinatora) [reply packages/transactions/src/create_transaction.ts:8]: Also I think it may be more readable and concise if you do the following
```suggestion
    const txNonce = typeof nonce === 'bigint' ? nonce :  BigInt(nonce);
    return new Transaction({ signerId, publicKey, nonce: txNonce, receiverId, actions, blockHash });
```

File: packages/utils/src/validators.ts
Comment (vikinatora) [inline packages/utils/src/validators.ts:37]: nitpick: better readablity
```suggestion
            currentSum = currentSum + (stakes[i] / mid);
```

File: packages/utils/src/validators.ts
Comment (vikinatora) [inline packages/utils/src/validators.ts:56]: Since ordering of bigints is used in two places already I think it's better to create a util file with a function that sorts it. It's will be much more understandable

**Review by vikinatora — APPROVED** [review]
Some nitpicks but looks very good

File: packages/utils/src/validators.ts
Comment (vikinatora) [inline packages/utils/src/validators.ts:24]: This is also a good candidate for a helper function.

**Review by vikinatora — DISMISSED** [review]

**Comment by gtsonevv** [comment]
Hey @gagdiez , this PR is ready for review. Merging is blocked until you approve it because you requested changes.

File: packages/utils/test/format.test.js
Comment (gagdiez) [inline packages/utils/test/format.test.js:62]: I see here we added the test for `NEAR_NOMINATION`, to make sure that the precision is right.

What puzzles me is that, just above this, there is a test checking that `formatNearAmount` is compatible with `parseNearAmount`. How is that it didn't fail when the bug of `BigInt(10**24)` was introduced?

They all need to be reviewed and, if necessary, more tests added. Otherwise, we cannot be sure we didn't added another bug to this cornerstone method.

**Comment by gagdiez** [comment]
@gtsonevv 

I am still puzzled about the tests, and why they didn't fail. Specially on the `format` file.

We need to find an explanation, and add more tests if necessary.

If I am being picky with this PR is because a rounding error could lead to real people losing real money.

File: packages/utils/test/format.test.js
Comment (gagdiez) [reply packages/utils/test/format.test.js:62]: ok, that explains it

**Review by gagdiez — APPROVED** [review]

**Comment by vikinatora** [comment]
Delaying the merge of this PR intentionally, because we want to batch these breaking changes with other oustanding breaking changes, in order to not increment major versions unnecessarily.

**Review by vikinatora — APPROVED** [review]

**Pull request merged**
